torch/distributed/_spmd/graph_optimization.py

```diff
@@ -443,3 +443,41 @@ def comm_fusion_with_concat(
     else:
         if begin < len(comm_blocks):
             _fuse_with_cat(gm, comm_blocks[begin:end], node_indices)
+
+
+@graph_optimization_pass(run_after=("comm_fusion_with_concat",))
+def schedule_comm_wait(gm: IterGraphModule) -> None:
+    """
+    Delay the execution of wait tensors of allreduce until its first user.
+    """
+    comm_blocks = get_all_comm_blocks(gm, (CommType.ALLREDUCE, "all_reduce"))
+
+    # Find all the end users.
+    allreduce_users = set()
+    for allreduce in comm_blocks:
+        for output in allreduce.outputs:
+            allreduce_users.update(output.users)
+
+    node_indices = {node: i for i, node in enumerate(gm.graph.nodes)}
+    for allreduce in comm_blocks:
+        # Find the earliest users.
+        assert (
+            len(allreduce.outputs) >= 1
+        ), f"Found a allreduce that has zero outputs/users -- {allreduce}."
+        # Initialize the target_node to be the first user of the first output.
+        target_node = next(iter(next(iter(allreduce.outputs)).users))
```

Comment:
> @mrshenli's internal review:  I guess this is assuming none of these users node will be reordered by subsequent graph optimization passes?

IMO, this is a general question for all optimization passes. If some subsequent optimization passes change the topological sort order, there is no guarantee that the previous optimization passes will still have the same effects, regardless the algorithm the previous passes adopt. I added `run_after` argument to the `graph_optimization_pass` contract to expect developers to consider the effect and add the order constraints to the graph optimization passes. But we may need a further discussion on this topic.